tests/parameter_search/test_automatic_rounding.py

```diff
@@ -0,0 +1,240 @@
+        "module__n_a_bits": n_bits,
+        "module__n_layers": 2,
+        "max_epochs": 10,
+    }
+
+    model = NeuralNetClassifier(**params)
+
+    x, y = load_data(NeuralNetClassifier)
+
+    model.fit(x, y)
+
+    adapter = PowerOfTwoScalingRoundPBSAdapter(model.quantized_module_)
+    round_pbs_patterns = adapter.process()
+    # Remove rounding in the network to perform inference without the optimization.
+    # We expect a network that was optimized with the power-of-two adapter
+    # to be exactly correct to the non-optimized one
+    for (_, node_op) in model.quantized_module_.quant_layers_dict.values():
+        if isinstance(node_op, QuantizedMixingOp):
+            node_op.rounding_threshold_bits = None
+            node_op.lsbs_to_remove = None
+
+    assert (
+        len(round_pbs_patterns) == 0
+    ), "Expected number of round PBS optimized patterns was not matched"
+    assert (
+        adapter.num_ignored_valid_patterns == 1
+    ), "Expected number of ignored round PBS optimizable patterns was not matched"
+
+    tlu_optimizer = TLUDeltaBasedOptimizer(
+        exactness=Exactness.APPROXIMATE,
+    )
+    cfg = Configuration(show_optimizer=False, additional_pre_processors=[tlu_optimizer])
+
+    # Compile the quantized model
+    model.compile(x, configuration=cfg, verbose=True)
+
+    # Find optimized TLUs
+    tlu_nodes = model.quantized_module_.fhe_circuit.graph.query_nodes(
+        custom_filter=lambda node: node.converted_to_table_lookup
+        and "opt_round_a" in node.properties["attributes"],
+        ordered=True,
+    )
+
+    # A single TLU should be present
+    assert len(tlu_nodes) == 1
+
+    count_reinterpret = 0
+    count_tlu = 0
+    for line in model.quantized_module_.fhe_circuit.mlir.split("\n"):
+        if "reinterpret_precision" in line:
+            regex = r"-> tensor<(\d+x)*\!FHE.[A-z]+<(\d+)"
+
+            matches = re.finditer(regex, line.strip())
+            for _, m in enumerate(matches, start=1):
+                bitwidth = int(m.group(2))
+                # If this is raising precision, ignore
+                if bitwidth > 24:
+                    continue
+                assert bitwidth <= (n_bits + 1), line
+                count_reinterpret += 1
+        if "apply_lookup_table" in line:
+            count_tlu += 1
+
+    assert count_reinterpret > 0, "Could not find reinterpret_cast nodes in graph to analyze"
+    assert count_tlu == 1, "This model should compile to an MLIR with a single PBS layer"
+
+
+@pytest.mark.parametrize("n_bits", range(4, 9))
+def test_tlu_optimization_cryptoparam_finding(load_data, n_bits):
+    """Tests that using approximate rounding the crypto-params optimizer is
+    not more constrained than with non-rounded execution.
+    """
+
+    params = {
+        "module__n_w_bits": n_bits,
+        "module__n_a_bits": n_bits,
+        "module__n_layers": 2,
+        "max_epochs": 10,
+    }
+
+    model = NeuralNetClassifier(**params)
```

Comment:
small detail but you could update `instantiate_model_generic` to be able to provide custom params and then use `preamble` , makes test easier to control